src/irr/asset/CGLILoader.cpp

```diff
@@ -0,0 +1,457 @@
+
+		asset::SAssetBundle CGLILoader::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
+		{
+			if (!_file)
+				return {};
+
+			const char* filename = _file->getFileName().c_str();
+			gli::texture texture = gli::load(filename);
+			if (texture.empty())
+			{
+				os::Printer::log("LOAD GLI: failed to load texture at path ", filename, ELL_ERROR);
+				return {};
+			}
+			
+		    const gli::gl glVersion(gli::gl::PROFILE_GL33);
+			const GLenum target = glVersion.translate(texture.target());
+			const auto format = getTranslatedGLIFormat(texture, glVersion);
+			IImage::E_TYPE imageType;
+			IImageView<ICPUImage>::E_TYPE imageViewType;
+
+			if (format.first == EF_UNKNOWN)
+			{
+				os::Printer::log("LOAD GLI: the file consist of not supported format", _file->getFileName().c_str(), ELL_ERROR);
+				return {};
+			}
+
+			switch (texture.target())
+			{
+				case gli::TARGET_1D:
+				{
+					imageType = IImage::ET_1D;
+					imageViewType = ICPUImageView::ET_1D;
+					break;
+				}
+				case gli::TARGET_1D_ARRAY:
+				{
+					imageType = IImage::ET_1D;
+					imageViewType = ICPUImageView::ET_1D_ARRAY;
+					break;
+				}
+				case gli::TARGET_2D:
+				{
+					imageType = IImage::ET_2D;
+					imageViewType = ICPUImageView::ET_2D;
+					break;
+				}
+				case gli::TARGET_2D_ARRAY:
+				{
+					imageType = IImage::ET_2D;
+					imageViewType = ICPUImageView::ET_2D_ARRAY;
+					break;
+				}
+				case gli::TARGET_3D:
+				{
+					imageType = IImage::ET_3D;
+					imageViewType = ICPUImageView::ET_3D;
+					break;
+				}
+				case gli::TARGET_CUBE:
+				{
+					imageType = IImage::ET_2D;
+					imageViewType = ICPUImageView::ET_CUBE_MAP;
+					break;
+				}
+				case gli::TARGET_CUBE_ARRAY:
+				{
+					imageType = IImage::ET_2D;
+					imageViewType = ICPUImageView::ET_CUBE_MAP_ARRAY;
+					break;
+				}
+				default:
+				{
+					imageViewType = ICPUImageView::ET_COUNT;
+					assert(0);
+					break;
+				}
+			}
+
+			ICPUImage::SCreationParams imageInfo;
+			imageInfo.format = format.first;
+			imageInfo.type = imageType;
+			imageInfo.flags = imageViewType == ICPUImageView::ET_CUBE_MAP_ARRAY ? ICPUImage::E_CREATE_FLAGS::ECF_CUBE_COMPATIBLE_BIT : static_cast<ICPUImage::E_CREATE_FLAGS>(0u);
+			imageInfo.samples = ICPUImage::ESCF_1_BIT;
+			imageInfo.extent.width = texture.extent().x;
+			imageInfo.extent.height = texture.extent().y;
+			imageInfo.extent.depth = texture.extent().z;
+			imageInfo.mipLevels = texture.levels();
+			imageInfo.arrayLayers = texture.faces() * texture.layers();
+
+			auto image = ICPUImage::create(std::move(imageInfo));
+
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texture.size());
```

Comment:
does `gli::texture::size()` include all the mipmaps etc?

Reply:
yes it does, it includes whole size